File: 03/main.py
LENGTH = 12

with open("input.txt") as f:
    data = f.read().strip().split("\n")

gamma = ""
epsilon = ""
for i in range(LENGTH):
    zeros = 0
    ones = 0
    for l in data:
        if l[i] == "1":
            ones += 1
        else:
            zeros += 1

    if ones > zeros:
        gamma += "1"
        epsilon += "0"
    else:
        gamma += "0"
        epsilon += "1"

print(int(gamma, 2) * int(epsilon, 2))

File: 03/main.cpp
#include "../common.h"
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    ifstream fp("input.txt");
    vector<string> lines = read_lines(fp);

    string gamma, epsilon;
    for (int i = 0; i < lines[0].length(); i++) {
        int ones = 0, zeros = 0;

        for (int j = 0; j < lines.size(); j++) {
            if (lines[j][i] == '0')
                zeros++;
            else
                ones++;
        }

        if (ones > zeros) {
            gamma += "1";
            epsilon += "0";
        } else {
            gamma += "0";
            epsilon += "1";
        }
    }

    cout << stoi(gamma, 0, 2) * stoi(epsilon, 0, 2) << "\n";
}